include/envoy/redis/codec.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+
+namespace Redis {
+
+/**
+ * All RESP types as defined here: https://redis.io/topics/protocol
+ */
+enum class RespType { Null, SimpleString, BulkString, Integer, Error, Array };
+
+/**
+ * A variant implementation of a RESP value optimized for peformance. A C++11 union is used for
+ * the underlying type so that no unnecessary allocations/constructions are needed.
+ */
+class RespValue {
+public:
+  RespValue() : type_(RespType::Null) {}
+  ~RespValue() { cleanup(); }
+
+  std::vector<RespValue>& asArray();
+  const std::vector<RespValue>& asArray() const;
+  std::string& asString();
+  const std::string& asString() const;
+  int64_t& asInteger();
+  int64_t asInteger() const;
+
+  RespType type() const { return type_; }
+  void type(RespType type);
+
+private:
+  union {
+    std::vector<RespValue> array_;
```

Comment:
Although union is more efficient, to keep it simple I'll avoid union here and just use three objects that will make destruction and construction super clean.